src/generic/fetch.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * Contains utility methods using 'fetch'.
+ */
+
+import {ApiError, ErrorCode} from '../core/api_error';
+import {BFSCallback} from '../core/file_system';
+
+/**
+ * Asynchronously download a file as a buffer or a JSON object.
+ * Note that the third function signature with a non-specialized type is
+ * invalid, but TypeScript requires it when you specialize string arguments to
+ * constants.
+ * @hidden
+ */
+export function fetchFileAsync(p: string, type: 'buffer', cb: BFSCallback<Buffer>): void;
+export function fetchFileAsync(p: string, type: 'json', cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void {
+  fetch(p)
+    .then((res) => {
+      if (!res.ok) {
+        return cb(new ApiError(res.status, "fetch error."));
```

Comment:
It shouldn't! This should only happen:

* If the file index is incorrect, and refers to a file that doesn't actually exist.
* If the network goes out, preventing network requests from working.

The unit tests do not test either of these occurrences. The existing error codes are nonsensical to any code written against the API. Most code running with BrowserFS is written in a different programming language/for the native environment (C/C++/Java/Node JS/etc) and expects the standard libc error codes.

In both cases, `EIO` seems like the correct error code to me.